client/src/client.rs

```diff
@@ -120,16 +120,26 @@ where
                 | StatusCode::UNAUTHORIZED
                 | StatusCode::FORBIDDEN
                 | StatusCode::NOT_FOUND
-                | StatusCode::UNPROCESSABLE_ENTITY => {
-                    let res = ValidationFail::decode_all(&mut resp.body().as_ref());
-                    let err = res.wrap_err(
-                        "Failed to decode error-response from Iroha. \
-                         You are likely using a version of the client library \
-                         that is incompatible with the version of the peer software",
-                    )?;
-                    Err(ClientQueryError::Validation(err))
-                }
-                _ => Err(ResponseReport::with_msg("Unexpected query response", resp).into()),
+                | StatusCode::UNPROCESSABLE_ENTITY => Err(ValidationFail::decode_all(
+                    &mut resp.body().as_ref(),
+                )
+                .map_or_else(
+                    |_| {
+                        ClientQueryError::Other(
+                            ResponseReport::with_msg("Query failed", resp)
+                                .map_or_else(
+                                    |_| eyre!(
+                                        "Failed to decode response from Iroha. \
+                                        Response is neither `ValidationFail` encoded value nor valid utf-8 string error response. \
+                                        You are likely using a version of the client library that is incompatible with the version of the peer software",
+                                    ),
+                                    Into::into
+                                ),
+                        )
+                    },
+                    ClientQueryError::Validation,
+                )),
+                _ => Err(ResponseReport::with_msg("Unexpected query response", resp).unwrap_or_else(core::convert::identity).into()),
```

Comment:
I don't think so, because this error indicate problems on iroha side (unhanded rejections for example) and they usually returned as strings, so we can return them without problem.